// @mastra/schema-compat@1.2.6/dist/chunk-UOL7NZ5L.cjs
'use strict';

var chunkFS3P4V5M_cjs = require('./chunk-FS3P4V5M.cjs');
var chunkROFMEFEF_cjs = require('./chunk-ROFMEFEF.cjs');
var chunkYIDCYW4I_cjs = require('./chunk-YIDCYW4I.cjs');
var z3 = require('zod/v3');
var module$1 = require('module');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var z3__default = /*#__PURE__*/_interopDefault(z3);

var SUPPORTED_TARGETS = /* @__PURE__ */ new Set(["draft-07", "draft-04", "draft-2020-12"]);
var ZOD_V4_TARGET_MAP = {
  "draft-07": "draft-7",
  "draft-04": "draft-4"
};
function convertToJsonSchema(zodSchema, options, adapterOptions) {
  const toJSONSchema = getToJSONSchema();
  if (!toJSONSchema) {
    throw new Error("z.toJSONSchema is not available. Ensure zod >= 3.25.0 is installed.");
  }
  const target = SUPPORTED_TARGETS.has(options.target) ? options.target : "draft-07";
  const jsonSchemaOptions = {
    target: ZOD_V4_TARGET_MAP[target] ?? target
  };
  if (adapterOptions.unrepresentable) {
    jsonSchemaOptions.unrepresentable = adapterOptions.unrepresentable;
  }
  if (adapterOptions.override) {
    jsonSchemaOptions.override = adapterOptions.override;
  }
  return toJSONSchema(zodSchema, jsonSchemaOptions);
}
var _toJSONSchema = null;
var _toJSONSchemaResolved = false;
var __require;
function getRequire() {
  if (!__require) {
    __require = module$1.createRequire((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === 'SCRIPT' && _documentCurrentScript.src || new URL('chunk-UOL7NZ5L.cjs', document.baseURI).href)));
  }
  return __require;
}
function pickToJSONSchema(mod) {
  const candidate = mod;
  const picks = [
    candidate?.toJSONSchema,
    candidate?.z?.toJSONSchema,
    candidate?.default?.toJSONSchema,
    candidate?.default?.z?.toJSONSchema
  ];
  const toJSONSchema = picks.find((fn) => typeof fn === "function");
  return toJSONSchema ?? null;
}
function resolveToJSONSchema(loadModule) {
  for (const moduleName of ["zod/v4", "zod"]) {
    try {
      const zodModule = loadModule(moduleName);
      const toJSONSchema = pickToJSONSchema(zodModule);
      if (toJSONSchema) {
        return toJSONSchema;
      }
    } catch {
    }
  }
  return null;
}
function getToJSONSchema() {
  if (_toJSONSchemaResolved) {
    return _toJSONSchema;
  }
  _toJSONSchema = resolveToJSONSchema((moduleName) => getRequire()(moduleName));
  _toJSONSchemaResolved = true;
  return _toJSONSchema;
}
function toStandardSchema4(zodSchema, adapterOptions = {}) {
  const wrapper = Object.create(zodSchema);
  const existingStandard = zodSchema["~standard"];
  const jsonSchemaConverter = {
    input: (options) => {
      return convertToJsonSchema(zodSchema, options, adapterOptions);
    },
    output: (options) => {
      return convertToJsonSchema(zodSchema, options, adapterOptions);
    }
  };
  Object.defineProperty(wrapper, "~standard", {
    value: {
      ...existingStandard,
      jsonSchema: jsonSchemaConverter
    },
    writable: false,
    enumerable: true,
    configurable: false
  });
  return wrapper;
}

// src/standard-schema/standard-schema.ts
function jsonSchemaOverride(ctx) {
  const zodSchema = ctx.zodSchema;
  if (ctx.jsonSchema.type === "object" && ctx.jsonSchema.properties !== void 0 && !ctx.jsonSchema.additionalProperties) {
    ctx.jsonSchema.additionalProperties = false;
  }
  if (zodSchema) {
    if (zodSchema?.type === "date") {
      ctx.jsonSchema.type = "string";
      ctx.jsonSchema.format = "date-time";
    } else if (zodSchema?.type === "object" && zodSchema._zod?.def?.catchall?.type === "unknown") {
      ctx.jsonSchema.additionalProperties = true;
    }
  }
  return void 0;
}
var JSON_SCHEMA_LIBRARY_OPTIONS = {
  unrepresentable: "any",
  override: jsonSchemaOverride
};
function isVercelSchema(schema) {
  return typeof schema === "object" && schema !== null && "_type" in schema && "jsonSchema" in schema && typeof schema.jsonSchema === "object";
}
function isZodV4(schema) {
  return typeof schema === "object" && schema !== null && "_zod" in schema;
}
function isZodV3(schema) {
  if (schema === null || typeof schema !== "object") {
    return false;
  }
  if (isZodV4(schema)) {
    return false;
  }
  if ("~standard" in schema) {
    const std = schema["~standard"];
    if (typeof std === "object" && std !== null && std.vendor === "zod" && !("jsonSchema" in std)) {
      return true;
    }
  }
  return schema instanceof z3__default.default.ZodType;
}
function toStandardSchema5(schema) {
  if (isStandardSchemaWithJSON(schema)) {
    return schema;
  }
  if (isZodV4(schema)) {
    return toStandardSchema4(schema, {
      unrepresentable: JSON_SCHEMA_LIBRARY_OPTIONS.unrepresentable,
      override: JSON_SCHEMA_LIBRARY_OPTIONS.override
    });
  }
  if (isZodV3(schema)) {
    return chunkYIDCYW4I_cjs.toStandardSchema(schema);
  }
  if (isVercelSchema(schema)) {
    return chunkFS3P4V5M_cjs.toStandardSchema(schema);
  }
  if (schema === null || typeof schema !== "object" && typeof schema !== "function") {
    throw new Error(`Unsupported schema type: ${typeof schema}`);
  }
  if (typeof schema === "function") {
    throw new Error(`Unsupported schema type: function (schema libraries should implement StandardSchemaWithJSON)`);
  }
  return chunkROFMEFEF_cjs.toStandardSchema(schema);
}
function isStandardSchema(value) {
  if (value === null || typeof value !== "object" && typeof value !== "function") {
    return false;
  }
  if (!("~standard" in value)) {
    return false;
  }
  const std = value["~standard"];
  return typeof std === "object" && std !== null && "version" in std && std.version === 1 && "vendor" in std && "validate" in std && typeof std.validate === "function";
}
function isStandardJSONSchema(value) {
  if (value === null || typeof value !== "object" && typeof value !== "function") {
    return false;
  }
  if (!("~standard" in value)) {
    return false;
  }
  const std = value["~standard"];
  if (typeof std !== "object" || std === null) {
    return false;
  }
  if (!("version" in std) || std.version !== 1 || !("vendor" in std)) {
    return false;
  }
  if (!("jsonSchema" in std) || typeof std.jsonSchema !== "object") {
    return false;
  }
  return typeof std.jsonSchema.input === "function" && typeof std.jsonSchema.output === "function";
}
function isStandardSchemaWithJSON(value) {
  return isStandardSchema(value) && isStandardJSONSchema(value);
}
function standardSchemaToJSONSchema(schema, options = {}) {
  const { target = "draft-07", io = "output", override = JSON_SCHEMA_LIBRARY_OPTIONS.override } = options;
  const jsonSchemaFn = schema["~standard"].jsonSchema[io];
  let jsonSchema = jsonSchemaFn({
    target,
    libraryOptions: {
      ...JSON_SCHEMA_LIBRARY_OPTIONS,
      override
    }
  });
  jsonSchema = JSON.parse(JSON.stringify(jsonSchema));
  return jsonSchema;
}

exports.JSON_SCHEMA_LIBRARY_OPTIONS = JSON_SCHEMA_LIBRARY_OPTIONS;
exports.isStandardJSONSchema = isStandardJSONSchema;
exports.isStandardSchema = isStandardSchema;
exports.isStandardSchemaWithJSON = isStandardSchemaWithJSON;
exports.standardSchemaToJSONSchema = standardSchemaToJSONSchema;
exports.toStandardSchema = toStandardSchema5;
