// @mastra/core@1.15.0/dist/chunk-OQDRPRKM.cjs
'use strict';

var EventEmitter = require('events');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var EventEmitter__default = /*#__PURE__*/_interopDefault(EventEmitter);

// src/events/pubsub.ts
var PubSub = class {
};
var EventEmitterPubSub = class extends PubSub {
  emitter;
  constructor(existingEmitter) {
    super();
    this.emitter = existingEmitter ?? new EventEmitter__default.default();
  }
  async publish(topic, event) {
    const id = crypto.randomUUID();
    const createdAt = /* @__PURE__ */ new Date();
    this.emitter.emit(topic, {
      ...event,
      id,
      createdAt
    });
  }
  async subscribe(topic, cb) {
    this.emitter.on(topic, cb);
  }
  async unsubscribe(topic, cb) {
    this.emitter.off(topic, cb);
  }
  async flush() {
  }
  /**
   * Clean up all listeners during graceful shutdown.
   */
  async close() {
    this.emitter.removeAllListeners();
  }
};

exports.EventEmitterPubSub = EventEmitterPubSub;
exports.PubSub = PubSub;
