// @mastra/core@1.15.0/dist/provider-registry-QXHI4DMY.cjs
'use strict';

var chunkXIQD3JK4_cjs = require('./chunk-XIQD3JK4.cjs');



Object.defineProperty(exports, "GatewayRegistry", {
  enumerable: true,
  get: function () { return chunkXIQD3JK4_cjs.GatewayRegistry; }
});
Object.defineProperty(exports, "PROVIDER_MODELS", {
  enumerable: true,
  get: function () { return chunkXIQD3JK4_cjs.PROVIDER_MODELS; }
});
Object.defineProperty(exports, "PROVIDER_REGISTRY", {
  enumerable: true,
  get: function () { return chunkXIQD3JK4_cjs.PROVIDER_REGISTRY; }
});
Object.defineProperty(exports, "getProviderConfig", {
  enumerable: true,
  get: function () { return chunkXIQD3JK4_cjs.getProviderConfig; }
});
Object.defineProperty(exports, "getRegisteredProviders", {
  enumerable: true,
  get: function () { return chunkXIQD3JK4_cjs.getRegisteredProviders; }
});
Object.defineProperty(exports, "isProviderRegistered", {
  enumerable: true,
  get: function () { return chunkXIQD3JK4_cjs.isProviderRegistered; }
});
Object.defineProperty(exports, "isValidModelId", {
  enumerable: true,
  get: function () { return chunkXIQD3JK4_cjs.isValidModelId; }
});
Object.defineProperty(exports, "parseModelString", {
  enumerable: true,
  get: function () { return chunkXIQD3JK4_cjs.parseModelString; }
});
